keps/sig-network/0006-20180314-coredns-GA-proposal.md

```diff
@@ -0,0 +1,132 @@
+* It is written in Go, making it memory-safe (kube-dns includes dnsmasq which is not). 
+* CoreDNS has [better performance](https://github.com/kubernetes/community/pull/1100#issuecomment-337747482) than [kube-dns](https://github.com/kubernetes/community/pull/1100#issuecomment-338329100) in terms of greater QPS, lower latency, and lower memory consumption. 
+* It supports a number of [use cases](https://coredns.io/2017/05/08/custom-dns-entries-for-kubernetes/) that kube-dns doesn't. As a general-purpose authoritative DNS server it has a lot of functionality that kube-dns couldn't reasonably be expected to add.
+
+### Goals
+
+* Have CoreDNS run as the default DNS server both after a new installation and an upgrade of the existing installation.
+* Make CoreDNS available as an [image in gcr.io](https://github.com/kubernetes/kubernetes/pull/59753) and define a workflow/process to update the CoreDNS versions in the future.
+* Provide a kube-dns to CoreDNS upgrade path with configuration translation.
+* Retain the configuration of a CoreDNS deployment from Alpha (1.9) or Beta (1.10).
+
+### Non-Goals
+
+* Translation of CoreDNS ConfigMap back to kube-dns (i.e., downgrade).
+* Retaining the ConfigMap of a custom/modified image of CoreDNS during upgrade. That is, if the user has modified the Corefile in the ConfigMap, those changes will not be retained (the user should back it up and redo the changes on the new ConfigMap).
+* Translation configuration of kube-dns to equivalent CoreDNS that is defined outside of the kube-dns ConfigMap. For example, modifications to the manifest or `dnsmasq` configuration.
+
+## Proposal
+
+The proposed solution is to enable the selection of CoreDNS as the default cluster service discovery DNS for Kubernetes, replacing kube-dns.
+As the cluster DNS service in Kubernetes is an add-on service that is deployed after Kubernetes itself, the deployment tools will comply to with Kubernetes guidelines and deploy CoreDNS instead of kube-dns.
+Some of the most used deployment tools have been upgraded by the CoreDNS team, in cooperation of the owners of these tools, to be able to deploy CoreDNS:
+* kubeadm
+* kube-up
+* minikube
+* kops
+
+For other tools, each maintainer would have to add the upgrade to CoreDNS.
+
+### Use Cases
+
+* CoreDNS supports all functionality of kube-dns and also addresses several use-cases kube-dns lacks. Some of the Use Cases are as follows: 
+    * Supporting [Autopath](https://coredns.io/plugins/autopath/), which reduces the high query load caused by the long DNS search path in Kubernetes
+    * Making an alias for an external name [#39792](https://github.com/kubernetes/kubernetes/issues/39792)
+    * Dynamically adding services to another domain, without running another server [#55](https://github.com/kubernetes/dns/issues/55)
+    * Adding an arbitrary entry inside the cluster domain (for example TXT entries [#38](https://github.com/kubernetes/dns/issues/38))
```

Comment:
I do not think we should document, endorse, or emphasize operations that do not go through Kubernetes API, except with the explicit warning that it is coredns specific and non-portable.

Reply:
Ok. So are you thinking we should pull these from the KEP?

Reply:
Maybe instead of listing the various extra features of CoreDNS, just link to a CoreDNS specific document. That way it's linked, but can be made clear it's out of scope for this KEP and k8s in general.